lib/Expression_gen.ml

```diff
@@ -1,262 +1,313 @@
+    [ ( Ast.PMinus
+      , fun ppf es ->
+          pp_unary ppf
+            (if is_scalar (List.hd_exn es) then "-%a" else "minus(%a)")
+            es )
+    ; (PPlus, fun ppf es -> pp_unary ppf "%a" es)
+    ; ( Transpose
+      , fun ppf es ->
+          pp_unary ppf
+            (if is_scalar (List.hd_exn es) then "transpose(%a)" else "%a")
+            es )
+    ; (PNot, fun ppf es -> pp_unary ppf "logial_negation(%a)" es)
+    ; ( Minus
+      , fun ppf es -> pp_scalar_binary ppf "(%a - %a)" "subtract(%a, %a)" es )
+    ; (Plus, fun ppf es -> pp_scalar_binary ppf "(%a + %a)" "add(%a, %a)" es)
+    ; ( Times
+      , fun ppf es -> pp_scalar_binary ppf "(%a * %a)" "multiply(%a, %a)" es )
+    ; ( Divide
+      , fun ppf es ->
+          if
+            is_matrix (second es)
+            && (is_matrix (first es) || is_row_vector (first es))
+          then pp_binary ppf "mdivide_right(%a, %a)" es
+          else pp_scalar_binary ppf "(%a / %a)" "divide(%a, %a)" es )
+    ; (Modulo, fun ppf es -> pp_binary ppf "modulus(%a, %a)" es)
+    ; (LDivide, fun ppf es -> pp_binary ppf "mdivide_left(%a, %a)" es)
+    ; ( EltTimes
+      , fun ppf es ->
+          pp_scalar_binary ppf "(%a * %a)" "elt_multiply(%a, %a)" es )
+    ; ( EltDivide
+      , fun ppf es -> pp_scalar_binary ppf "(%a / %a)" "elt_divide(%a, %a)" es
+      )
+    ; (Pow, fun ppf es -> pp_binary ppf "pow(%a, %a)" es)
+    ; (Equals, fun ppf es -> pp_binary ppf "logical_eq(%a, %a)" es)
+    ; (NEquals, fun ppf es -> pp_binary ppf "logical_neq(%a, %a)" es)
+    ; (Less, fun ppf es -> pp_binary ppf "logical_lt(%a, %a)" es)
+    ; (Leq, fun ppf es -> pp_binary ppf "logical_lte(%a, %a)" es)
+    ; (Greater, fun ppf es -> pp_binary ppf "logical_gt(%a, %a)" es)
+    ; (Geq, fun ppf es -> pp_binary ppf "logical_gte(%a, %a)" es) ]
+    |> List.map ~f:(fun (k, v) -> (Operators.operator_name k, v))
+    |> Map.Poly.of_alist_exn
+  in
+  let misc_special_math_fns =
+    Map.Poly.of_alist_exn
+      [ ("lmultiply", fun ppf es -> pp_binary ppf "multiply_log(%a, %a)" es)
+      ; ( "lchoose"
+        , fun ppf es -> pp_binary ppf "binomial_coefficient_log(%a, %a)" es )
+      ; ("target", fun ppf _ -> pf ppf "get_lp(lp__, lp_accum__)")
+      ; ("get_lp", fun ppf _ -> pf ppf "get_lp(lp__, lp_accum__)")
+      ; ( "max"
+        , fun ppf es ->
+            if List.length es = 2 then pp_binary ppf "std::max(%a, %a)" es
+            else pp_ordinary_fn ppf f es )
+      ; ( "min"
+        , fun ppf es ->
+            if List.length es = 2 then pp_binary ppf "std::min(%a, %a)" es
+            else pp_ordinary_fn ppf f es )
+      ; ( "ceil"
+        , fun ppf es ->
+            if is_scalar (first es) then pp_unary ppf "std::ceil(%a)" es
+            else pp_ordinary_fn ppf f es ) ]
+  in
+  let read_data_or_param ppf es =
+    let i_or_r =
+      match ut with
+      | Ast.UInt -> "i"
+      | UReal -> "r"
+      | UVector | URowVector | UMatrix | UArray _
+       |UFun (_, _)
```

Comment:
Fat-fingered the last review too soon.  Is there something meaningful about the inconsistent punctuation around `|` here?  I mainly want syntactic consistency so things don't jump out at me as being different.

Reply:
Yep! It's a continuation of the previous cases - so UFun has the same action as UArray, UMatrix, etc.

Reply:
Ah, thanks---missed the lack of `->` in the previous row.  Now I'm glad it punctuates it weirdly so I'll at least notice.